File: src/python/xerxes_agent/channels/adapters/feishu.py
# Copyright 2025 The EasyDeL/Xerxes Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License")
"""Feishu / Lark adapter (ByteDance enterprise IM)."""

from __future__ import annotations

import json
import typing as tp

from .._helpers import WebhookChannel, http_post, parse_json_body
from ..types import ChannelMessage, MessageDirection


class FeishuChannel(WebhookChannel):
    """Feishu (Lark) bot adapter via app webhook + open-API send.

    Expects a tenant access token to be supplied directly or via an
    ``token_provider`` callable that returns the current token (the
    Feishu token expires every 2 h; the provider lets the caller cache
    + refresh out-of-band).
    """

    name = "feishu"

    def __init__(
        self,
        tenant_access_token: str = "",
        *,
        token_provider: tp.Callable[[], str] | None = None,
        api_base: str = "https://open.feishu.cn",
        http_client: tp.Any = None,
    ) -> None:
        """Configure the adapter with a static or refreshable tenant token.

        Args:
            tenant_access_token: Static tenant access token (used when no
                ``token_provider`` is supplied).
            token_provider: Callable returning the current tenant token.
            api_base: Feishu / Lark open-API base URL.
            http_client: Optional injected HTTP callable for tests.
        """
        super().__init__()
        self.tenant_access_token = tenant_access_token
        self.token_provider = token_provider
        self.api_base = api_base.rstrip("/")
        self._http = http_client

    def _resolve_token(self) -> str:
        """Return the current tenant token, preferring ``token_provider``."""
        if self.token_provider is not None:
            tok = self.token_provider()
            if tok:
                return tok
        return self.tenant_access_token

    def _parse_inbound(self, headers, body):
        """Decode a Feishu event envelope into a :class:`ChannelMessage`.

        Drops ``url_verification`` handshakes. Feishu nests the text body
        as a JSON-encoded string under ``event.message.content``; it is
        parsed here to recover the plain ``text`` field.
        """
        data = parse_json_body(body)
        if not data:
            return []
        if data.get("type") == "url_verification":
            return []
        ev = data.get("event") or {}
        message = ev.get("message") or {}
        sender = ev.get("sender") or {}
        text = ""
        try:
            content = json.loads(message.get("content", "{}"))
            text = content.get("text", "")
        except Exception:
            text = message.get("content", "")
        if not text:
            return []
        return [
            ChannelMessage(
                text=text,
                channel=self.name,
                channel_user_id=sender.get("sender_id", {}).get("open_id", ""),
                room_id=message.get("chat_id", ""),
                platform_message_id=message.get("message_id", ""),
                direction=MessageDirection.INBOUND,
            )
        ]

    async def _send_outbound(self, message):
        """POST a text message to the IM open API addressed by ``chat_id``."""
        url = f"{self.api_base}/open-apis/im/v1/messages?receive_id_type=chat_id"
        body = {
            "receive_id": message.room_id,
            "msg_type": "text",
            "content": json.dumps({"text": message.text}),
        }
        http_post(
            url,
            json_body=body,
            headers={"Authorization": f"Bearer {self._resolve_token()}"},
            http_client=self._http,
        )

File: src/typescript/src/channels/feishu.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { postJson, providerUrl, type ChannelFetch } from './http.js'
import {
  recordValue,
  resolveAccessToken,
  stringValue,
  throwOnProviderErrorCode,
  urlWithQuery,
  type ChannelAccessTokenProvider,
  type RelayChannelTransport,
} from './relay.js'
import { createChannelMessage, MessageDirection, type ChannelMessage } from './types.js'
import {
  parseJsonBody,
  WebhookChannel,
  type WebhookHeaders,
  type WebhookResponse,
} from './webhooks.js'

const FEISHU_API_BASE = 'https://open.feishu.cn/'

/** Feishu/Lark event callbacks are relayed here; encrypted callbacks are not decrypted. */
export const FEISHU_TRANSPORT: RelayChannelTransport = {
  inbound: 'webhook-relay',
  outbound: 'http-api',
  unsupported: ['encrypted event payload decryption', 'persistent WebSocket event delivery'],
}

export interface FeishuChannelOptions {
  /** Override for Lark-compatible deployments or tests. */
  readonly apiBaseUrl?: string
  readonly fetchImplementation?: ChannelFetch
  readonly tenantAccessToken?: string
  /** Called on every outbound request to support externally refreshed tokens. */
  readonly tokenProvider?: ChannelAccessTokenProvider
}

/**
 * Feishu/Lark webhook relay and `im/v1/messages` text sender.
 *
 * URL-verification challenges are answered, while encrypted webhook events
 * and long-lived provider connections intentionally remain outside this
 * dependency-free adapter.
 */
export class FeishuChannel extends WebhookChannel {
  readonly name = 'feishu'
  readonly transport = FEISHU_TRANSPORT

  private readonly apiBaseUrl: string
  private readonly fetchImplementation: ChannelFetch | undefined
  private readonly staticToken: string
  private readonly tokenProvider: ChannelAccessTokenProvider | undefined

  constructor(options: FeishuChannelOptions = {}) {
    super()
    this.apiBaseUrl = options.apiBaseUrl ?? FEISHU_API_BASE
    this.fetchImplementation = options.fetchImplementation
    this.staticToken = options.tenantAccessToken ?? ''
    this.tokenProvider = options.tokenProvider
  }

  override async handleWebhook(
    headers: WebhookHeaders,
    body: Uint8Array,
  ): Promise<WebhookResponse> {
    const payload = parseJsonBody(body)
    if (payload.type === 'url_verification') {
      return {
        status: 200,
        body: stringValue(payload.challenge),
        headers: { 'content-type': 'text/plain; charset=utf-8' },
      }
    }
    return super.handleWebhook(headers, body)
  }

  protected parseInbound(
    _headers: WebhookHeaders,
    body: Uint8Array,
  ): readonly ChannelMessage[] {
    const payload = parseJsonBody(body)
    if (payload.type === 'url_verification') {
      return []
    }
    const event = recordValue(payload.event)
    const message = recordValue(event.message)
    const sender = recordValue(event.sender)
    const senderId = recordValue(sender.sender_id)
    const text = feishuText(message.content)
    if (!text) {
      return []
    }
    return [createChannelMessage({
      channel: this.name,
      direction: MessageDirection.INBOUND,
      text,
      channelUserId: stringValue(senderId.open_id),
      roomId: stringValue(message.chat_id),
      platformMessageId: stringValue(message.message_id),
      metadata: { message_type: stringValue(message.message_type) },
    })]
  }

  protected async sendOutbound(message: ChannelMessage): Promise<void> {
    if (!message.roomId) {
      throw new TypeError('Feishu outbound messages require roomId')
    }
    const token = await resolveAccessToken(this.staticToken, this.tokenProvider, 'Feishu')
    const url = urlWithQuery(providerUrl(this.apiBaseUrl, 'open-apis/im/v1/messages'), 'receive_id_type', 'chat_id')
    const response = await postJson(url, {
      body: {
        receive_id: message.roomId,
        msg_type: 'text',
        content: JSON.stringify({ text: message.text }),
      },
      headers: { Authorization: `Bearer ${token}` },
      ...(this.fetchImplementation ? { fetchImplementation: this.fetchImplementation } : {}),
    })
    throwOnProviderErrorCode(response, 'Feishu', 'code')
  }
}

function feishuText(value: unknown): string {
  const raw = stringValue(value)
  if (!raw) {
    return ''
  }
  try {
    return stringValue(recordValue(JSON.parse(raw)).text)
  } catch {
    return raw
  }
}
